rust/ccommon_rs/src/log.rs

```diff
@@ -0,0 +1,414 @@
+}
+
+#[repr(u32)]
+#[derive(Debug, PartialEq, PartialOrd, Eq)]
+pub enum LoggerStatus {
+    OK = 0,
+    LoggerNotSetupError = 1,
+    RegistrationFailure = 2,
+    LoggerAlreadySetError = 3,
+    InvalidUTF8 = 4,
+}
+
+impl From<LoggingError> for LoggerStatus {
+    fn from(e: LoggingError) -> Self {
+        match e {
+            LoggingError::LoggerRegistrationFailure => LoggerStatus::RegistrationFailure,
+            LoggingError::LoggingAlreadySetUp => LoggerStatus::LoggerAlreadySetError,
+        }
+    }
+}
+
+/// This function will set up our logger as the default
+/// one for the `log` crate at the given
+/// `level`. This function must be called as early
+/// as possible in program setup, followed by
+/// a call to [`log_rs_set`]
+///
+/// [`log_rs_set`]: fn.log_rs_set.html
+///
+/// # Errors
+///
+/// If we fail to set up our logger, we will print a
+/// message on stderr and return
+/// [`LoggerStatus::RegistrationFailure`], which means
+/// we could not register ourselves as the provider
+/// of the logging backend for the `log` crate.
+/// This should be treated as a fatal error because
+/// one cannot un-register the existing backend, and
+/// this operation will *never* succeed.
+///
+/// If this method had been called previously,
+/// and we are the provider of the logging framework,
+/// we return [`Ok`].
+///
+/// # Safety
+///
+/// The caller must ensure that the lifetime of `logger`
+/// lives until `rust_cc_log_destroy`
+/// is called or the program terminates.
+#[no_mangle]
+pub extern "C" fn log_rs_setup() -> LoggerStatus {
+    match try_init_logger() {
+        Ok(_) => LoggerStatus::OK,
+        Err(LoggingError::LoggingAlreadySetUp) => {
+            eprintln!("cc_log logging already set up");
+            LoggerStatus::OK
+        }
+        Err(LoggingError::LoggerRegistrationFailure) => {
+            eprintln!("Error setting up cc_log! {}", LoggingError::LoggerRegistrationFailure);
+            LoggerStatus::RegistrationFailure
+        }
+    }
+}
+
+/// This function sets the cc_log logger instance to be the
+/// sink for messages logged from the `log` crate. The user
+/// must call [`log_rs_setup`] _before_ calling this function
+/// to register us as the backend for the `log` crate.
+///
+/// # Panics
+///
+/// This function will panic if the `logger` pointer is NULL.
+///
+/// # Errors
+///
+/// Returns [`LoggerNotSetupError`] if [`log_rs_setup`] was NOT
+/// called prior to this function being called.
+///
+/// If there's already been a `logger` instance set up, then we will return
+/// [`LoggerAlreadySetError`]. This error need not be fatal.
+///
+/// [`log_rs_setup`]: fn.log_rs_setup.html
+/// [`LoggerNotSetupError`]: enum.LoggerStatus.html
+/// [`LoggerAlreadySetError`]: enums.LoggerStatus.html
+///
+/// # Undefined Behavior
+///
+/// If the `logger` pointer becomes invalid before [`log_rs_unset`] is called, the
+/// behavior is undefined.
+///
+/// [`log_rs_unset`]: fn.log_rs_unset.html
```

Comment:
I lean toward either naming the rs-C APIs `rs_*` (e.g. `rs_log_setup()`) or `*_rs`(`log_setup_rs()`) even just for the sake of simpler regex for search.